### PR TITLE
[Toolbar] Implement a public API for adding toolbars

### PR DESCRIPTION
**Pull request by psarram**
Implement a toolbar registry and a toolbar plugin to provide a toolbar for a selected object.

Use the toolbar API to add a layout provider which defines the toolbar structure for objects when selected inside a display layout. 

To group controls in a toolbar, the idea is to add a divider in the toolbar structure like any other control:

```
{
     "control: "divider"
}
```

The layout toolbar doesn't use a divider and so grouping is not implemented yet.

### CODE REVIEW

File: platform/commonUI/edit/src/representers/EditToolbar.js [hunk 1]
```diff
@@ -156,23 +144,17 @@ define(
...
-                var count = 0;
-                // Show/hide each item
-                (section.items || []).forEach(function (item) {
-                    item.hidden = !self.isApplicable(item);
-                    count += item.hidden ? 0 : 1;
-                });
-                // Hide this section if there are no applicable items
-                section.hidden = !count;
+            // Show/hide controls per applicability
+            function refreshApplicability(item) {
```
Comment (larkin) [inline platform/commonUI/edit/src/representers/EditToolbar.js]: Applicability should be handled in the toolbar provider's `toolbar` method: it should only return controls that are valid for a given selection.

**Comment by psarram** [comment]
@larkin I've addressed your initial comments in the latest changes. Please review when you get a chance.

File: platform/commonUI/edit/test/representers/EditToolbarSpec.js [hunk 2]
```diff
@@ -40,182 +41,107 @@ define(
...
-                expect(state[structure.sections[0].items[0].key])
-                    .toEqual(testABC.a);
-                expect(state[structure.sections[0].items[1].key])
-                    .toEqual(testABC.b);
-                expect(state[structure.sections[0].items[2].key])
-                    .toEqual(testABC.c);
-            });
-
-            it("reads properties from getters", function () {
+            xit("reads properties from getters", function () {
```
Comment (akhenry) [inline platform/commonUI/edit/test/representers/EditToolbarSpec.js]: We should make sure that there are no disabled tests before merge.

File: platform/features/layout/bundle.js [hunk 3]
```diff
@@ -305,6 +283,27 @@ define([
...
+                {
+                    name: "Display Layout Toolbar",
+                    key: "layout",
+                    description: "A toolbar for objects inside a display layout.",
+                    forSelection: function (selection) {
+                        // Apply the layout toolbar if the selected object is inside a layout.
+                        return (selection && selection[1] && selection[1].context.item.type === 'layout');
+                    },
+                    toolbar: function (selection) {
+                        return [
```
Comment (akhenry) [inline platform/features/layout/bundle.js]: This is still very much tied to the old API, which we will need to revisit during refactoring. 

I wonder if perhaps we should take an approach similar to the [IndicatorsAPI]() here where either a simple control with some customizable options can be provided (based on however our custom form controls end up being defined), or an arbitrary DOM element can be provided.

Comment (larkin) [reply platform/features/layout/bundle.js]: This is a legacy patch that allows us to refactor toolbars without having to convert bundles to plugins. The “public” side of this is exposed separate from the legacy api. 

The toolbar api is still draft after this merges, and I expect some additions to support custom controls, but that should be tackled during our forms overhaul.

**Comment by psarram** [comment]
**Author Checklist**

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y - unit tests were updated with changes.
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

**Comment by psarram** [comment]
Fixes #1908 and #1917

**Comment by psarram** [comment]
@akhenry, @larkin Thanks for the review! I turned the preprocessors back on and repaired all the tests. The xit tests are deleted since they did not relate to the new logic. New tests need to be written once the upgraded test framework is in place.

**Comment by akhenry** [comment]
## Reviewer Checklist

* Changes appear to address issue? Y
* Appropriate unit tests included? Y
* Code style and in-line documentation are appropriate? Y
* Commit messages meet standards? Y

**Pull request merged**